File: Algorithms/Python3.x/921-Minimum_Add_to_Make_Parentheses_Valid.py
# Runtime: 32 ms, faster than 94.89% of Python3 online submissions for Minimum Add to Make Parentheses Valid.
# Memory Usage: 13.8 MB, less than 8.33% of Python3 online submissions for Minimum Add to Make Parentheses Valid.
class Solution:
    def minAddToMakeValid(self, S: str) -> int:
        left_brackets = 0
        min_add = 0
        for s in S:
            if s == '(':
                left_brackets += 1
            elif s == ')':
                left_brackets -= 1
            if left_brackets < 0:
                min_add += 0 - left_brackets
                left_brackets = 0
        return min_add + left_brackets

File: Algorithms/C++/921-Minimum_Add_to_Make_Parentheses_Valid.cpp
// Runtime: 4 ms, faster than 58.46% of C++ online submissions for Minimum Add to Make Parentheses Valid.
// Memory Usage: 8.4 MB, less than 84.62% of C++ online submissions for Minimum Add to Make Parentheses Valid.
class Solution {
public:
    int minAddToMakeValid(string S) {
        int left_brackets = 0,
            min_add = 0;
        for (char s : S) {
            if (s == '(')
                ++left_brackets;
            else
                --left_brackets;
            if (left_brackets < 0) {
                min_add += 0 - left_brackets;
                left_brackets = 0;
            }
        }
        return min_add + left_brackets;
    }
};
